crates/data_components/src/s3_vectors/vector_table.rs

```diff
@@ -476,19 +514,123 @@ impl S3VectorsTable {
+                                return Err(Error::MaxSpillAttemptsReached);
+                            }
+                        }
+                    } else {
+                        return Err(Error::S3VectorPutVectorError {
+                            source: service_error.into_err(),
+                        });
+                    }
+                }
+                Err(e) => {
+                    return Err(Error::S3VectorPutVectorError {
+                        source: e.into_service_error(),
+                    });
+                }
+            }
         }
+    }
 
-        tracing::info!(
-            "S3 Vectors Index {index_name} updated; records={records}, duration={duration:?}",
-            index_name = self.idx,
-            records = vectors.len(),
-            duration = start.elapsed()
-        );
+    fn is_capacity_exceeded_error(error: &PutVectorsError) -> bool {
+        matches!(error, PutVectorsError::ServiceQuotaExceededException(_))
+    }
+
+    async fn get_next_spill_index(
+        &self,
+        current_index: &S3VectorIdentifier,
+    ) -> Result<Option<S3VectorIdentifier>> {
+        let (bucket_name, base_index_name) = match current_index {
+            S3VectorIdentifier::Index {
+                bucket_name,
+                index_name,
+            } => {
+                if SpillIndex::is_spill_index(index_name) {
+                    match SpillIndex::parse(index_name) {
+                        Ok(Some(spill)) => (bucket_name.clone(), spill.base_name),
+                        _ => return Ok(None), // Invalid spill index format
+                    }
+                } else {
+                    // This is the main index
+                    (bucket_name.clone(), index_name.clone())
+                }
+            }
+            S3VectorIdentifier::IndexArn(_) => {
+                // ARN-based indexes don't support spilling
+                return Ok(None);
+            }
+        };
 
+        // Start from 01 for main index, or continue from current spill index
+        let mut spill_num = if let Some(index_name) = current_index.index_name() {
+            if SpillIndex::is_spill_index(index_name) {
+                match SpillIndex::parse(index_name) {
+                    Ok(Some(spill)) => spill.sequence + 1,
+                    _ => 1,
+                }
+            } else {
+                1
+            }
+        } else {
+            1
+        };
+
+        loop {
+            if spill_num > 99 {
+                return Ok(None);
+            }
+            let spill_name = format!("{base_index_name}.{spill_num:02}");
+            match self.create_spill_index(&bucket_name, &spill_name).await {
+                Ok(()) => {
+                    return Ok(Some(S3VectorIdentifier::Index {
+                        bucket_name: bucket_name.clone(),
+                        index_name: spill_name,
+                    }));
+                }
+                Err(_) => {
```

Comment:
Is the idea that if the spill index is already created we should skip it? It might be better to explicitly check for the Conflict in that case. Or if the index is already created, we could try to use it?